fist/server.c

```diff
@@ -1,142 +1,194 @@
-            } else {
-                dstring document = commands.values[1];
-                dstring text = djoin(drange(commands, 2, commands.length), ' ');
-                printf("TEXT: '%s'\n", dtext(text));
-                dstringa index = indexer(text, 10);
-                printf("INDEX SIZE: %d\n", index.length);
-                for(int i = 0; i < index.length; i++) {
-                    dstring on = index.values[i];
-                    hm = hset(hm, on, document);
-                }
-                send(new_socket, INDEXED, strlen(INDEXED), 0);
-            }
-        } else if(dequals(commands.values[0], dcreate(SEARCH))) {
-            printf("SEARCH\n");
-            if(commands.length < 2) {
-                send(new_socket, TOO_FEW_ARGUMENTS, strlen(TOO_FEW_ARGUMENTS), 0);
-            } else {
-                dstring text = djoin(drange(commands, 1, commands.length), ' ');
-                printf("SEARCH STRING: '%s'\n", dtext(text));
-                dstringa value = hget(hm, text);
-                printf("SEARCH RESULTS SIZE: %d\n", value.length);
-                if(!value.length) {
-                    send(new_socket, NOT_FOUND, strlen(NOT_FOUND), 0);
-                } else {
-                    dstring output = dcreate("[");
-                    for(int i = 0; i < value.length; i++) { // This builds the JSON array output
-                        dstring on = value.values[i];
-                        output = dappendc(output, '"');
-                        output = dappendd(output, on);
-                        output = dappendc(output, '"');
-                        if(i != value.length - 1)
-                            output = dappendc(output, ',');
-                    }
-                    output = dappendc(output, ']');
-                    output = dappendc(output, '\n');
-                    send(new_socket, dtext(output), output.length, 0);
-                }
-            }
-        } else {
-            send(new_socket, INVALID_COMMAND, strlen(INVALID_COMMAND), 0);
-        }
+  // TODO: don't allocate new dstrings all the time
+  if (dequals(trimmed, dcreate(EXIT))) {
+    // TODO: don't calculate strlen all the time
+    send(fd, BYE, strlen(BYE), 0);
+    return 1;
+  }
+  if (dequals(commands.values[0], dcreate(INDEX))) {
+    printf("INDEX\n");
+    if (commands.length < 3) {
+      send(fd, TOO_FEW_ARGUMENTS, strlen(TOO_FEW_ARGUMENTS), 0);
+    } else {
+      dstring document = commands.values[1];
+      dstring text = djoin(drange(commands, 2, commands.length), ' ');
+      printf("TEXT: '%s'\n", dtext(text));
+      dstringa index = indexer(text, MAX_PHRASE_LENGTH);
+      printf("INDEX SIZE: %d\n", index.length);
+      for (int i = 0; i < index.length; i++) {
+	dstring on = index.values[i];
+	hm = hset(hm, on, document);
+      }
+      send(fd, INDEXED, strlen(INDEXED), 0);
     }
-    close(new_socket);
-    sdump(hm); // Save new DB state to file
```

Comment:
After the changes, `sdump()` is never called.

I wasn't sure where to do it, previously it was whenever a client disconnected. We could:

* Add a save command
* Save periodically on a configurable timer
* Save when the server exits
* Or do something else